doc/adr/0008-jetstream-kv.md

```diff
@@ -0,0 +1,319 @@
+we use the new `Nats-Expected-Last-Subject-Sequence` header.
+
+#### Retrieving Values
+
+There are different situations where messages will be retrieved using different APIs, below describes the different models.
+
+In all cases we return a generic `Entry` type.
+
+Deleted data - (see later section on deletes) - has the `KV-Operation` header set to `DEL`, a value received from either of these
+methods with this header set indicates the data was being deleted.  If this is a pending=0 message it means the key does not exist.
+
+##### Get Operation
+
+We have extended the `io.nats.jetstream.api.v1.stream_msg_get_request` API to support loading the latest value for a specific
+subject.  Thus a read for `CONFIGURATION.username` becomes a `io.nats.jetstream.api.v1.stream_msg_get_request` with the
+`last_by_subject` set to `$KV.CONFIGURATION.username`.
+
+##### History and Watch Operations
+
+These operations requires access to all values for a key, to achieve this we create an ephemeral consumer on filtered by the subject
+and read the entire value list in using `deliver_all`.
+
+JetStream will report the Pending count for each message, the latest value from the available history would have a pending of `0`.
+When constructing historic values, dumping all values etc we ensure to only return pending 0 messages as the final value
+
+#### Deleting Values
+
+Since the store supports history - via the `max_age` for messages - we should preserve history when deleting keys. To do this we
+place a new message in the subject for the key with a nil body and the header `KV-Operation: DEL`.
+
+This preserves history and communicate to watchers, caches and gets that a delete operation should be handled - clear cache,
+return no key error etc.
+
+#### Purging a bucket
+
+A purge deletes all messages from a bucket, this is a normal stream purge on the entire stream.
+
+#### Deleting a bucket
+
+Remove the stream entirely
+
+#### Watchers
+
+Watchers support sending received `PUT` and `DEL` operations across a channel or language specific equivelant.
+
+##### Key Watch
+
+This is a consumer set up to get the last value for a key. The watch remains active till cancelled and any PUT/DEL on this key should be 
+made available to the client.
+
+##### Bucket Watch
+
+A bucket watch is a consumer set up on the wildcard subject for the bucket - `$KV.<Bucket Name>.*` - and sending every single message
+over the channel to the client.
+
+This means historical values will be sent as well as latest values. In this case we cannot provide the delta on a per key basis. But
+the delta will represent the remainder for the entire bucket, this allows Watcher clients like the memory cache to set themselves
+ready once the full history was read.
+
+When a stream is empty at create time - `NumPending+Delivered.Consumer==0` - a nil is sent across the channel to signal to the
+caller that they should not expect any values.
+
```

Comment:
Do values contain an indication to wether the current value is currently the last one int the stream?
smth like seqno==max_seqno
Knowing when to stop would allow Watch to be used to iterate.

Reply:
if you look at `Delta` - it has the last value on the key when 0 (this is on history). For watchers, a key watcher only gets the last value, and then notifications from each new modification. For bucket watchers, you get all values and all modifications (this is why a JetStream consumer option of LastValuePerSubject would be very useful).

Reply:
I read the delta as per key. not as current last.

Reply:
The delta is per key yes.  We dont know the key delta on a bucket watch though. But in practise thats fine because the watch is used to build a consistent local view of the KV - so once I am at pending=0 for the bucket it implies delta for all keys are 0.